### PR TITLE
Feature OVF improvement

### PR DESCRIPTION
**Pull request by smavros**
- Capitalization ignored from `Filter_File_Handler` both for `istringstreams` and `keyword` arguments
- OVF text file format can be read in
- New function `Read_ColumnVector_Configuration()`. In the future it should be extended to be used by the functions which read the SPIRIT file format.
- There is no implementation for reading SPIRIT_REGULAR_POS and SPIRIT_CSV yet.
- `IO_Chain_Read()` is not tested yet for any format.

### CODE REVIEW

**Comment by smavros** [comment]
@GPMueller this PR can be merged after review.

File: core/src/io/Dataparser.cpp [hunk 1]
```diff
@@ -1059,32 +1070,31 @@ namespace IO
...
             Log( lvl, sender, fmt::format( "# OVF data representation = {}", ovf_data_representation ) );
             Log( lvl, sender, fmt::format( "# OVF binary length       = {}", ovf_binary_length ) );
             
             // Check that representation and binary length valures are ok
-            if( ovf_data_representation != "Text" && ovf_data_representation != "Binary" )
+            if( ovf_data_representation != "text" && ovf_data_representation != "binary" )
             {
                 spirit_throw(Utility::Exception_Classifier::Bad_File_Content, Utility::Log_Level::Error,
-                    "Data representation must be either \"Text\" or \"Binary\"");
+                    "Data representation must be either \"text\" or \"binary\"");
```
Comment (GPMueller) [inline core/src/io/Dataparser.cpp]: This is not right - the keywords *should* be in uppercase, i.e. "Text" or "Binary", according to how MuMax writes OVF files.

File: core/src/io/Dataparser.cpp [hunk 2]
```diff
@@ -1059,32 +1070,31 @@ namespace IO
             myfile.Read_String( ovf_data_representation, "# Begin: Data" );
             std::istringstream repr( ovf_data_representation );
             repr >> ovf_data_representation;
-            if( ovf_data_representation == "Binary" ) 
+            if( ovf_data_representation == "binary" )
```
Comment (GPMueller) [inline core/src/io/Dataparser.cpp]: I believe this is incorrect, as the `ovf_binary_length` is not read in via the `Filter_File_Handle::Read_String` function. As it is therefore not lowercased, this will nor compare correctly.
An alternative that would be more readable in code would be to write a function `IO::lowercase(std::string & str)`, the comparison could then look like `if( lowercase(ovf_data_representation) == "binary")`.

File: core/test/test_io.cpp [hunk 3]
```diff
@@ -40,7 +40,6 @@ TEST_CASE( "IO", "[io]" )
         // set config to minus z and write the system out
         Configuration_MinusZ( state.get() );
         IO_Image_Write( state.get(), filename, filetype, "io test" );
-        IO_Image_Append( state.get(), filename, filetype, "io test" );
```
Comment (GPMueller) [inline core/test/test_io.cpp]: Why did you remove `IO_Image_Append`, `IO_Chain_Write` and `IO_Chain_Append`?

File: core/src/io/Filter_File_Handle.cpp [hunk 4]
```diff
@@ -139,9 +143,10 @@ namespace IO
...
 
-    void Filter_File_Handle::Read_String( std::string& var, const std::string keyword, 
-                                            bool log_notfound )
+    void Filter_File_Handle::Read_String( std::string& var, std::string keyword, bool log_notfound )
     {
+        std::transform( keyword.begin(), keyword.end(), keyword.begin(), ::tolower );
+        
         if ( Find( keyword ) )
         {
             getline( this->iss, var );
```
Comment (GPMueller) [inline core/src/io/Filter_File_Handle.cpp]: This will not lowercase the read-in string, `var` should also be transformed after this.

File: core/src/io/Dataparser.cpp [hunk 5]
```diff
@@ -1059,32 +1070,31 @@ namespace IO
...
+                    "Data representation must be either \"text\" or \"binary\"");
             }
             
-            if( ovf_data_representation == "Binary" && 
+            if( ovf_data_representation == "binary" && 
                  ovf_binary_length != 4 && ovf_binary_length != 8  )
             {
                 spirit_throw(Utility::Exception_Classifier::Bad_File_Content, Utility::Log_Level::Error,
-                    "Binary representation can be either \"Binary 8\" or \"Binary 4\"");
+                    "Binary representation can be either \"binary 8\" or \"binary 4\"");
```
Comment (GPMueller) [inline core/src/io/Dataparser.cpp]: In all these logging cases, it should be "Binary" and "Text" (i.e. first letter uppercase), as this is the way that MuMax writes it's files.

File: core/src/io/Dataparser.cpp (hunk 2 above)
Comment (GPMueller) [inline core/src/io/Dataparser.cpp]: I believe the above will not lowercase the entire line and will therefore not lowercase `ovf_data_representation`. It may be that this way MuMax files cannot be read in anymore, as they use "Data Binary" and "Data Text".

Comment (smavros) [reply core/src/io/Dataparser.cpp]: The `ovf_data_representation` is read, together with the keyword `"Begin: Data"`, by the `Read_String()` method of `Filter_File_Handle` object. At first the line under investigation is de-capitalized by `Filter_File_Handle::GetLine()` function. Additionally, the `keyword` argument is de-capitalized by `Read_String()` to assure that the keyword will matched in the line no matter how the `keyword` was formatted in the call of the function. For that reason the `ovf_data_representation` variable can only be lower case. 

If we change in the code the comparison string `"binary"` to `"Binary"` the `Read_From_OVF()` will fail. It would make sense to do the inverse of what you are proposing eg `ovf_data_representation == lowercase("Binary")` but of course one can just simply write `"binary"` instead.

Comment (smavros) [reply core/src/io/Dataparser.cpp]: MuMax would be read in without a problem since `Filter_File_Handler::GetLine()` will convert every line to the header to lower case. That case will work like that:
- The whole line is de-capitalized by `Filter_File_Handler::GetLine()`
- In that line we are searching for the string `"# Begin: Data"` which of course will be de-capitalized by `Read_String()` otherwise it would never match to the already de-capitalized line.
- There are 3 possible states for the variable `ovf_data_representation`. Those are `"text"`, `"binary 4"` or `"binary 8"`. In case of binary we want to discard the binary length.
- We initialize the `std::istringstream repr` with `ovf_data_representation` variable and we are keeping only the first word. ( `>>` would stop in `" " `).

File: core/src/io/Dataparser.cpp (hunk 1 above)
Comment (smavros) [reply core/src/io/Dataparser.cpp]: MuMax is not following the official OVF 2.0 format described in NIST page ( see http://math.nist.gov/oommf/doc/userguide12a5/userguide/OVF_2.0_format.html ). Of course by applying de-capitalization in functions `GetLine()`, `Read_Single()` and `Read_String()` we solve the problem of capitalization and we expect MuMax to be parsed in without any problems.

File: core/src/io/Dataparser.cpp (hunk 5 above)
Comment (smavros) [reply core/src/io/Dataparser.cpp]: We are following the official OVF 2.0 format. Since we are dealing with capitalization properly this exception will be raised only in case of misspelling of the data type field eg `"txt"` instead of `"text"`in case that someone tries to parse in a human-written OVF file. Of course the choice is up to you.

File: core/src/io/Filter_File_Handle.cpp (hunk 4 above)
Comment (smavros) [reply core/src/io/Filter_File_Handle.cpp]: - The attribute of the `Filter_File_Handle` object `std::string line` is populated by the function `Filter_File_Handle::GetLine_Handle()` for each line of the file. In case that the whole line is not a comment, `line` is processed by a call in `Filter_File_Handle::GetLine()`. 
- The function `Filter_File_Handle::GetLine()` is __de-capitalizing__ every `line`.
- The `std::istringstream iss` is cleared and populated for each `line` of the input file by the call to  `Filter_File_Handle::Find_in_Line()`.
- We are getting `var` from attribute `iss`.
Hence the `var` will always hold a decapitalized string.

**Comment by smavros** [comment]
To ensure that an OVF file will be parsed in regardless of its capitalization we added a test function in `test_io.cpp` which takes the, generated by the test, `"image_ovf_text.ovf"` file converts it to upper case, writes it out and then tries to read it in again. This test will ensure the correctness of text capitalization handling by `IO::Filter_File_Handle` in the case of future changes.

File: core/src/io/Dataparser.cpp [hunk 6]
```diff
@@ -1071,32 +1082,31 @@ namespace IO
             myfile.Read_String( ovf_data_representation, "# Begin: Data" );
             std::istringstream repr( ovf_data_representation );
             repr >> ovf_data_representation;
```
Comment (GPMueller) [inline core/src/io/Dataparser.cpp]: One last question: why these two lines?
```
std::istringstream repr( ovf_data_representation );
repr >> ovf_data_representation;
```
It seems to me like extracting the first word of a sequence, but the contents of `ovf_data_representation` should already be just "binary" or "text" right?

Comment (smavros) [reply core/src/io/Dataparser.cpp]: Actually what we are doing is:

- Before the call to `Read_String()` the `ovf_data_representation` is just an empty string. 
- The `Read_String()` will get whatever is on the line, after the `keyword` (in our case `"# Begin: Data"`), till `\n`.
- The trick is that now `ovf_data_represantation` may __not__ hold only the data representation because in case of binary representation we would get `"binary x"` (where `x` is 4 or 8) instead of only `"binary"`. 
- We turn the result which is stored in `ovf_data_represantation` to an `std::istringstream` and we stream its first word to `ovf_data_represantation`.
- If `ovf_data_represantation` is `"text"` the `std::istringstream` is empty and we are done. If it is `"binary"` then the `std::istringstream` is just holding the binary length that we are going to stream in to the variable `ovf_binary_length`.
